Admin/Test-AMSI.ps1

```diff
@@ -1,144 +1,891 @@
+        param(
+            [Parameter(Mandatory = $true)]
+            [string]$ExchangeServer
+        )
+
+        Write-Host ""
+        Write-Host "AMSI Providers detection:" -ForegroundColor Green
+
+        $AMSIProvidersSB = {
+            $AMSIProviders = Get-ChildItem 'HKLM:\SOFTWARE\Microsoft\AMSI\Providers' -Recurse -ErrorAction SilentlyContinue
+            if ($AMSIProviders) {
+                Write-Host "Providers:"
+                $AMSIProviders.Name | Out-Host
+                $providerCount = 0
+                foreach ($provider in $AMSIProviders) {
+                    $provider -match '[0-9A-Fa-f\-]{36}' | Out-Null
+                    foreach ($m in $Matches.Values) {
+                        $key = "HKLM:\SOFTWARE\Classes\ClSid\{$m}"
+                        $providers = $null
+                        $providers = Get-ChildItem $key -ErrorAction SilentlyContinue
+                        if ($providers) {
+                            $providerCount++
+                            Write-Host ""
+                            Write-Host "Provider $($providerCount):"
+                            $providers | Format-Table -AutoSize | Out-Host
+                            $path = $null
+                            $path = ($providers | Where-Object { $_.PSChildName -eq 'InprocServer32' }).GetValue('')
+                            if ($path) {
+                                $WindowsDefenderPath = $path.Substring(1, $path.LastIndexOf("\"))
+                                if ($WindowsDefenderPath -like '*Windows Defender*') {
+                                    Write-Host "Windows Defender with AMSI integration found." -ForegroundColor Green
+                                    $checkCmdLet = $null
+                                    $checkCmdLet = Get-Command Get-MpComputerStatus -ErrorAction SilentlyContinue
+                                    if ($null -eq $checkCmdLet) {
+                                        Write-Warning "Get-MpComputerStatus cmdLet is not available"
+                                    } else {
+                                        if ((Get-MpComputerStatus).RealTimeProtectionEnabled) {
+                                            Write-Host "Windows Defender has Real Time Protection Enabled" -ForegroundColor Green
+                                        } else {
+                                            Write-Warning "Windows Defender has Real Time Protection Disabled"
+                                        }
+                                    }
+                                    Write-Host "It should be version 1.1.18300.4 or newest."
+                                    if (Test-Path $WindowsDefenderPath -PathType Container) {
+                                        $folder = Get-ChildItem  $WindowsDefenderPath | Sort-Object LastWriteTime -Descending | Select-Object -First 1
+                                        $process = Join-Path $folder.FullName "MpCmdRun.exe"
```

Comment:
This doesn't make sense here. Why are we listing out all the child items of `$WindowsDefenderPath` which is a directory, then sorting the objects by last write time and just assuming that is a folder? This doesn't look right.

Reply:
uhm, I think you are right.
I found a better way to do this with the information in the registry key